Feature/dkg integration

hsaleemsupra opened the pull request:
See https://github.com/Entropy-Foundation/smr-moonshot/issues/2183.

hsaleemsupra committed:
update validator pub keys to support dkg

hsaleemsupra committed:
update validator pub keys to support dkg

hsaleemsupra committed:
adds blockmetadata flag

hsaleemsupra committed:
adds blockmetadata flag

hsaleemsupra committed:
featureflag for blockmetadataext

hsaleemsupra committed:
featureflag for blockmetadataext

hsaleemsupra committed:
add consenses_key support for past version

hsaleemsupra committed:
add consenses_key support for past version

hsaleemsupra committed:
add clan, family committee structure

hsaleemsupra committed:
add clan, family committee structure

hsaleemsupra committed:
update dkg module to handle new dkg

hsaleemsupra committed:
update dkg module to handle new dkg

hsaleemsupra committed:
change dkg.move to use DkgNodeConfig

hsaleemsupra committed:
change dkg.move to use DkgNodeConfig

hsaleemsupra committed:
on going changes

hsaleemsupra committed:
on going changes

hsaleemsupra committed:
moved dkg transaction validation to aptosvm

hsaleemsupra committed:
moved dkg transaction validation to aptosvm

hsaleemsupra committed:
adds dkg transaction to post bls pk shares

hsaleemsupra committed:
adds dkg transaction to post bls pk shares

hsaleemsupra committed:
removed author from dkgmeta

hsaleemsupra committed:
removed author from dkgmeta

hsaleemsupra committed:
supra dkg feature flag

hsaleemsupra committed:
supra dkg feature flag

hsaleemsupra committed:
rebase with dev

hsaleemsupra committed:
Merge branch 'feature/dkg-integration' of https://github.com/Entropy-Foundation/aptos-core into feature/dkg-integration

hsaleemsupra committed:
add dkg validator transaction validation

hsaleemsupra committed:
adds family check in dkg transaction validation

hsaleemsupra committed:
minor

hsaleemsupra committed:
adds feature gating for validator transaction execution

hsaleemsupra committed:
update dkg transaction execution validation

isaacdoidge committed:
[#2300] Feature flag.

hsaleemsupra committed:
temporarily support old consensus key

hsaleemsupra committed:
added identity to dkg node config

hsaleemsupra committed:
added dkg events

hsaleemsupra committed:
add consensus key rust type

hsaleemsupra committed:
add dkg event try from for Contract event

hsaleemsupra committed:
changes to get dkg move state

hsaleemsupra committed:
added DKGResharing struct

hsaleemsupra committed:
emit dkg update events

hsaleemsupra committed:
impl from dkg events to contract event

hsaleemsupra committed:
to_bytes for consensuskey

hsaleemsupra committed:
update crypto version

hsaleemsupra committed:
update crypto

hsaleemsupra committed:
update crypto

hsaleemsupra committed:
update crypto

hsaleemsupra committed:
update crypto

hsaleemsupra committed:
update crypto

hsaleemsupra committed:
validator key change and refactor

hsaleemsupra committed:
Revert "validator key change and refactor"

This reverts commit 3d14a9897f085d037698e8e3bf0dbdf9f47a4684.

hsaleemsupra committed:
updated validator public keys and refactoring

isaacdoidge commented on consensus/src/epoch_manager.rs:
```diff
@@ -952,7 +952,7 @@ impl<P: OnChainConfigProvider> EpochManager<P> {
             dkg_session.transcript.as_slice(),
         )
-        .map_err(NoRandomnessReason::TranscriptDeserializationError)?;
+            .map_err(NoRandomnessReason::TranscriptDeserializationError)?;
```
What formatter are you using?

hsaleemsupra replied:
I will apply the formatter in the end. Didn't want to many changes in the code that become hard to review.

isaacdoidge commented on aptos-move/framework/supra-stdlib/sources/class_groups.move:
```diff
@@ -0,0 +1,47 @@
+module supra_std::class_groups {
```
Need to add copyright statements for Supra.

hsaleemsupra replied:
The move modules donot seem to have copyright statements. Only the native function files have copyright statements. I will add a statement in natives file. Do we need the statement in move module as well?

hsaleemsupra replied:
fixed

isaacdoidge commented on aptos-move/framework/supra-framework/sources/reconfiguration_with_dkg.move:
```diff
@@ -34,14 +36,21 @@ module supra_framework::reconfiguration_with_dkg {
         };
         reconfiguration_state::on_reconfig_start();
         let cur_epoch = reconfiguration::current_epoch();
+        let randomness_seed = randomness::bytes(32);
```
What inspired the change from `randomness_config::current()` to `randomness::bytes(32)` and what difference does it make?

hsaleemsupra replied:
randomness_config::current() was used by aptos's dkg to provide the validators with some config information related to randomness used in their protocol e.g. **secrecy_threshold** and **reconstruction_threshold**. However our dkg protocol doesn't require that information. Rather it only requires a random 32 byte seed which should be unbiasable and unpredictable in each epoch. The assumption here is that we can get such a seed by using the randomness module with the function **randomness::bytes(32)**.

isaacdoidge commented on aptos-move/framework/supra-framework/sources/dkg.spec.move:
```diff
@@ -13,15 +13,22 @@ spec supra_framework::dkg {
     }
 
-    spec finish(transcript: vector<u8>) {
+    spec set_dkg_meta(dkg_meta_all_committees: vector<u8>) {
```
Nice work updating the spec :)

isaacdoidge commented on aptos-move/framework/aptos-stdlib/sources/any.move:
```diff
@@ -26,6 +26,14 @@ module aptos_std::any {
     }
 
+    /// Create a new `Any` struct with provided parameters
+    public fun new(type_name: String, data: vector<u8>): Any {
```
Is `pack` not suitable for the context that you're using `new` in? I haven't reached it yet so perhaps it will be obvious when I get there.

hsaleemsupra replied:
The `new` method is already part of the dev branch but current branch is not synced with dev. We cannot use pack here since pack internally performs serialization of data using `to_bytes`. With `new` method the assumption is we already have serialized data available from somewhere and we wish to deserialize it only using `unpack` later.

isaacdoidge commented on aptos-move/aptos-vm/src/validator_txns/mod.rs:
```diff
@@ -19,15 +19,20 @@ impl AptosVM {
             ValidatorTransaction::ObservedJWKUpdate(jwk_update) => {
                 self.process_jwk_update(resolver, log_context, session_id, jwk_update)
             },
+            _ => Err(VMStatus::Error {
```
Are there no other variants to match or is it expected that the other variants should never reach here? If there are no other variants then it is better not to add the `_` branch so that the compiler will indicate that this code needs to be updated when a new variant is added. However, if the variants are expected not to reach here then `UNREACHABLE` does not seem like a good choice of error code as a bug could result in the code being executed.

hsaleemsupra replied:
We do have another variant of ValidatorTransaction i.e. [DKGResult](https://github.com/Entropy-Foundation/aptos-core/blob/213ca23970599c4a7e3b9070502b808b3f49fa84/types/src/validator_txn.rs#L14). This variant is used by the aptos dkg code. One option is to remove this variant. However for that we will have to remove the aptos dkg related code as well since it uses this transaction type internally. However aptos dkg code touches a large part of the codebase and removing it requires significant refactoring effort which may be done separately. 

For now we know that the DKGResult variant would not be called as our dkg doesn't uses that variant. What do you suggest should be done in that case? Should I change the error variant for this case?

isaacdoidge replied:
Alright, yes, good point about removing the old DKG code separately. The current status is fine, but it would be better to explicitly match `ValidatorTransaction::DKGResult` instead of using `_` so that we know to update the code if/when we add additional variants of `ValidatorTransaction`.

hsaleemsupra replied:
fixed

isaacdoidge commented on api/types/src/transaction.rs:
```diff
@@ -865,6 +882,50 @@ impl From<DKGTranscript> for ExportedDKGTranscript {
+pub struct DKGTransaction {
+    #[serde(flatten)]
+    #[oai(flatten)]
+    pub info: TransactionInfo,
```
Fields should only be `pub` if they are expected to be writable outside the defining module. This is generally not the case with transactions. Is there a reason why these fields should be `pub`?

hsaleemsupra replied:
fixed

isaacdoidge commented on api/types/src/transaction.rs:
```diff
@@ -865,6 +882,50 @@ impl From<DKGTranscript> for ExportedDKGTranscript {
 
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Object)]
+pub struct ExportedDKGTransactionData {
+    pub epoch: U64,
```
Same question here about `pub` access.

hsaleemsupra replied:
The fields information is needed by function `convert_validator_transaction` in `aptos_indexer_grpc_fullnode::convert`.

isaacdoidge commented on api/types/src/transaction.rs:
```diff
@@ -865,6 +882,50 @@ impl From<DKGTranscript> for ExportedDKGTranscript {
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Object)]
+pub struct ExportedDKGTransactionData {
```
Please add `///` comments to describe each of the fields and why they are necessary.

hsaleemsupra replied:
fixed

isaacdoidge commented on api/types/src/transaction.rs:
```diff
@@ -865,6 +882,50 @@ impl From<DKGTranscript> for ExportedDKGTranscript {
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Object)]
+pub struct ExportedDKGTransactionData {
+    pub epoch: U64,
+    pub author: Address,
```
I find that ordering struct fields alphabetically by name makes code easier to read. Up to you though.

isaacdoidge commented on api/types/src/transaction.rs:
```diff
@@ -865,6 +882,50 @@ impl From<DKGTranscript> for ExportedDKGTranscript {
+    pub bls_aggregate_signature: Vec<u8>,
+    pub signer_indices_clan_committee: Vec<u32>,
+    pub transaction_type: u8,
+    pub payload: HexEncodedBytes,
```
It seems good to also add the chain id as a field of this struct as an extra measure to prevent replaying across instances of the protocol.

hsaleemsupra replied:
chain_id is part of the smr_transaction and is verified when the transaction is added for execution. Do we need to store chain_id info on chain as well?

isaacdoidge replied:
Actually, adding the chain id to this struct won't help much. What matters is whether the `bls_aggregate_signature` (by way of the transcript, which is what the signature appears to be on) factors it in or not. If it doesn't and the same committee runs two parallel chains (an unlikely but not unreasonable scenario) then a Byzantine node from that committee could replay DKG transactions from one chain on the other. In general, all signatures that are generated for a specific chain must cover the chain id.

hsaleemsupra replied:
That is a good point. We can add chain id as part of the domain string used when creating the signatures. I will make that change.

isaacdoidge replied:
Can you please point me to where you've made the change?

isaacdoidge replied:
Please share this code if you've made the change.

hsaleemsupra replied:
For the DKGMeta transaction, the signed data is of type DKGMetaWithAggregateSignatureAllCommittees which contains a vector of DKGMetaCompressed for various committees and each DKGMetaCompressed contains EpochId which also contains ChainId. 
Similarly for Public Shares transaction, we sign data of type AggregateCommitmentAllCommittees which contains a vector of type AggregateCommitment each containing EpochId.

isaacdoidge replied:
Great, thanks for explaining.

isaacdoidge commented on aptos-move/aptos-vm/src/aptos_vm.rs:
```diff
@@ -2630,6 +2633,10 @@ impl AptosVM {
                 (VMStatus::Executed, output)
             },
             Transaction::ValidatorTransaction(txn) => {
+                if !self.features().is_enabled(FeatureFlag::SUPRA_DKG) {
```
Might be better to put this inside the DKG branch of `process_validator_transaction`.

hsaleemsupra replied:
fixed

isaacdoidge commented on types/src/on_chain_config/aptos_features.rs:
```diff
@@ -94,6 +94,7 @@ pub enum FeatureFlag {
     PRIVATE_POLL = 91,
     SUPRA_AUTOMATION_TASK_SYNC = 92,
     SUPRA_COUNT_FAILED_PROPOSALS = 93,
+    SUPRA_DKG = 94,
```
We should add this to the default features as well when the feature is finished and fully integrated into `smr_moonshot`.

hsaleemsupra replied:
Will add a todo

isaacdoidge commented on types/src/dkg/transactions.rs:
```diff
@@ -0,0 +1,71 @@
+
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq, CryptoHasher, BCSCryptoHash)]
+pub struct DKGTransactionMetadata {
+    pub epoch: u64,
```
Same comments r.e. `pub`. Please revisit all new structs and determine if they actually need to be `pub`. Private should be the default.

hsaleemsupra replied:
The struct fields are used by aptos_vm's `validate_dkg_validator_transaction` method to validate the dkg transaction.

isaacdoidge replied:
If you add the `derive_getters` [crate](https://docs.rs/derive-getters/latest/derive_getters/) and import `derive_getters::Getters`, then add it to the `#[derive(` macro then you will have access to auto-generated getter methods that return immutable refs for the fields. Read-only objects should only expose immutable refs.

hsaleemsupra replied:
fixed

isaacdoidge commented on types/src/dkg/transactions.rs:
```diff
@@ -0,0 +1,71 @@
+        }
+    }
+
+    pub fn dummy() -> Self {
```
If this only used in test code? If so, it would be good to guard it with conditional compilation.

hsaleemsupra replied:
fixed

isaacdoidge reviewed (commented):
Some initial comments. I still have most of the code left to review.

isaacdoidge commented on aptos-move/aptos-vm/src/aptos_vm.rs:
```diff
@@ -2819,6 +2826,119 @@ impl VMValidator for AptosVM {
+            },
+        }
+
+        // the node submitting the transaction must be a family node
```
Do we need this requirement? Wouldn't it be safe for any node to submit the transaction, as long as the clan signature is valid?

hsaleemsupra replied:
I will remove this requirement. If the clan committee signature is valid the transaction should be considered valid.

isaacdoidge commented on aptos-move/aptos-vm/src/validator_txns/dkg.rs:
```diff
@@ -5,132 +5,108 @@ use crate::{
+        ) {
             Ok((vm_status, vm_output)) => Ok((vm_status, vm_output)),
-            Err(Expected(failure)) => {
-                // Pretend we are inside Move, and expected failures are like Move aborts.
```
We need to do this too. The VM should return an error status code for the transaction but continue operation when an expected failure occurs. If the VM emits an actual `Err` then execution of the whole block will fail and the consensus will panic. Are you catching the new errors anywhere else? Perhaps I've not got there yet.

hsaleemsupra replied:
fixed

isaacdoidge commented on aptos-move/aptos-vm/src/validator_txns/dkg.rs:
```diff
@@ -5,132 +5,108 @@ use crate::{
-        )
-        .map_err(|_| Expected(TranscriptDeserializationFailed))?;
+        let function_name;
+        let args;
```
We don't tend to use pre-initialization in Rust. See my suggestion below for a more idiomatic approach.

hsaleemsupra replied:
fixed

isaacdoidge commented on aptos-move/aptos-vm/src/validator_txns/dkg.rs:
```diff
@@ -5,132 +5,108 @@ use crate::{
+                    dkg_transaction.data_bytes.as_move_value(),
+                ];
+            },
+        }
```
I've suggested this in GH so the formatting will be off. Hopefully the idea is clear though.
```suggestion
        let (function_name, args) = match dkg_transaction.metadata.transaction_type {
            DKGTransactionType::DKGMeta => (SET_DKG_META, vec![dkg_transaction.data_bytes.as_move_value()]),
            DKGTransactionType::PublicKeyShares => (FINISH_WITH_DKG_RESULT,  vec![
                    MoveValue::Signer(AccountAddress::ONE),
                    dkg_transaction.data_bytes.as_move_value(),
                ]),
        }
```

hsaleemsupra replied:
fixed

isaacdoidge commented on aptos-move/framework/src/natives/cryptography/class_groups.rs:
```diff
@@ -0,0 +1,73 @@
+    debug_assert!(arguments.len() == 1);
+
+    //todo: update gas cost
+    context.charge(BLS12381_BASE)?;
```
We will need to increment the gas version in `aptos-move/aptos-gas-schedule/src/ver.rs` and define a corresponding new gas parameter for CG validation in `aptos-move/aptos-gas-schedule/src/gas_schedule/aptos_framework.rs` under the new gas version. This will be sufficient to feature-gate the new functions.

isaacdoidge replied:
Actually, I see that you've defined the new module inside `supra_stdlib`, so we should create a new gas parameters type for that library like there is for `move_stdlib` (see `aptos-move/aptos-gas-schedule/src/gas_schedule/move_stdlib.rs`) and define the new parameter there instead of in `aptos_framework.rs`.

hsaleemsupra replied:
fixed

isaacdoidge commented on aptos-move/framework/supra-stdlib/sources/class_groups.move:
```diff
@@ -0,0 +1,47 @@
+    use std::option;
+    use std::option::Option;
+
+    struct CGPublicKey has copy, drop, store {
```
Do you think that it might be helpful to differentiate between validated and unvalidated public keys? See `ed25519.move` for an example.

hsaleemsupra replied:
I do not currently see any use for that. We can do that if need arises in the future.

isaacdoidge commented on aptos-move/aptos-vm/src/aptos_vm.rs:
```diff
@@ -2819,6 +2826,119 @@ impl VMValidator for AptosVM {
+            ValidatorTransaction::DKG(txn) => txn,
+            _ => {
+                return VMValidatorResult::error(
+                    StatusCode::ONLY_DKG_TRANSACTION_VALIDATION_SUPPORTED,
```
All of the errors emitted by this function except for this one are expected under certain system configurations or when some validators are Byzantine. This error is an internal error that can only be triggered if a bug is introduced to the code. Are there any other examples of internal errors being mapped to VM errors like this?

hsaleemsupra replied:
I think it might be better to just pass `DKGTransactionData` in the function: `validate_dkg_validator_transaction` as it only expects a dkg validator transaction. In that case we would not have this issue where someone calls this function with an invalid `ValidatorTransaction` variant.

hsaleemsupra replied:
fixed

isaacdoidge commented on types/src/dkg/dkg_committee.rs:
```diff
@@ -0,0 +1,24 @@
+/// Reflection of `0x1::dkg_committee::DkgNodeConfig` in rust.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct DkgNodeConfig {
+    pub addr: AccountAddress,
```
It would be helpful to be more descriptive here.
```suggestion
    pub pool_address: AccountAddress,
```

hsaleemsupra replied:
fixed

isaacdoidge commented on types/src/dkg/mod.rs:
```diff
@@ -1,219 +1,74 @@
+    if let Some(family_node_indices) = family_committee_indices {
+        let result = family_node_indices
+            .iter()
+            .any(|x| dealer_committee.committee[*x].addr == addr);
```
Do you ensure that `dealer_committee.committee.len() == family_node_indices.len()` anywhere? If not, then this deref is unsafe.

hsaleemsupra replied:
`family_node_indices` is a subset of `dealer_committee` so this should be safe. However, we do not need this now as we want to remove the check that the transaction submitter is a family node.

isaacdoidge commented on types/src/dkg/mod.rs:
```diff
@@ -1,219 +1,74 @@
-            },
-            transcript_bytes: vec![],
+        for signer in signers {
+            let clan_node_index = clan_committee_indices[*signer as usize];
```
Have you checked the bounds somewhere?

hsaleemsupra replied:
I will replace this with `get()` just to be safe.

hsaleemsupra replied:
fixed

isaacdoidge reviewed (commented):
A few more comments. Still a few files left to cover.

hsaleemsupra committed:
ongoing changes

hsaleemsupra committed:
add gas params for supra_stdlib

hsaleemsupra committed:
fmt

isaacdoidge commented on aptos-move/framework/src/natives/cryptography/class_groups.rs:
```diff
@@ -0,0 +1,78 @@
+    debug_assert!(_ty_args.is_empty());
+    debug_assert!(arguments.len() == 1);
+
+    context.charge(CLASS_GROUPS_PER_PUBKEY_DESERIALIZE * NumArgs::one())?;
```
Dr Joshi's recommendation for tuning new gas parameters is to implement them in Move then create a tx targeting the related function, and apply a discount to its `gas_used` value. This seems like a reasonable approximation to use until we have the processes and tools required to measure precisely. How much effort do you think that it would take to estimate these new parameters in that way?

hsaleemsupra replied:
This seems like a significant effort for class group key deserialization function. This would involve implementing the [zk proof of possession](https://github.com/Entropy-Foundation/crypto/blob/dkg_integration/src/bls12381/key_pop_zk.rs#L13) for class group elements as well as implementing the [deserialization logic for the class group elements](https://github.com/Entropy-Foundation/bicycl-rs/blob/dev/src/cl/hsm_utils.rs#L201) from the bicycl library in Move.

isaacdoidge commented on aptos-move/aptos-gas-schedule/src/gas_schedule/supra_stdlib.rs:
```diff
@@ -0,0 +1,19 @@
+    NativeGasParameters => .supra_stdlib,
+    [
+        // Note(Gas): this initial value is guesswork.
+        [class_groups_per_pubkey_deserialize: InternalGasPerArg, "class.groups.per_pubkey_deserialize", 400684],
```
We need to increment [the gas schedule version](https://github.com/Entropy-Foundation/aptos-core/blob/dev/aptos-move/aptos-gas-schedule/src/ver.rs#L91) and ensure that these entries are only included in that version or greater (see [here](https://github.com/Entropy-Foundation/aptos-core/blob/dev/aptos-move/aptos-gas-schedule/src/gas_schedule/aptos_framework.rs#L337) for an example).

hsaleemsupra replied:
I have added the tag `RELEASE_V1_16_SUPRA_V1_8_0` for the new version. Is that okay?

isaacdoidge commented on aptos-move/aptos-gas-schedule/src/gas_schedule/supra_stdlib.rs:
```diff
@@ -0,0 +1,19 @@
+crate::gas_schedule::macros::define_gas_parameters!(
+    SupraStdlibGasParameters,
+    "supra_stdlib",
+    NativeGasParameters => .supra_stdlib,
```
We will also need to test to see how the addition of this new schedule affects nodes running the old binary after the framework has been upgraded. Areg might be able to help with this when the feature is complete.

hsaleemsupra replied:
Noted

isaacdoidge replied:
Please document this in the PR description so that we don't forget.

isaacdoidge replied:
This is still TODO.

isaacdoidge commented on api/types/src/transaction.rs:
```diff
@@ -741,6 +748,15 @@ impl
+                    dkg_transaction_data: dkg_transaction_data.into(),
+                })
+            },
+
```
```suggestion
```

isaacdoidge commented on api/types/src/transaction.rs:
```diff
@@ -749,6 +765,7 @@ impl
                     dkg_transcript: dkg_transcript.into(),
                 })
             },
+
```
```suggestion
```

isaacdoidge commented on types/src/validator_txn.rs:
```diff
@@ -2,29 +2,31 @@
 #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize, CryptoHasher, BCSCryptoHash)]
 pub enum ValidatorTransaction {
+    DKG(DKGTransactionData),
     DKGResult(DKGTranscript),
```
Do we still need to retain the old `DKGResult` variant?

hsaleemsupra replied:
The variant is used by the legacy dkg code by aptos. It is being used in `DKGManager` used by aptos dkg. We can may be remove the aptos dkg code in a separate refactor.

isaacdoidge commented on aptos-move/aptos-vm/src/aptos_vm.rs:
```diff
@@ -2819,6 +2824,102 @@ impl VMValidator for AptosVM {
+        };
+
+        if agg_sig
+            .verify_aggregate_arbitrary_msg(&[dkg_transaction.data_bytes.as_slice()], &[&agg_pk])
```
How large are these signatures? We sign message digests in the consensus rather than the full messages to reduce the size of the signature.

hsaleemsupra replied:
The signature size is independent of the msg size. Since we are using bls min_pk, the signature size is always 96 bytes. Also the signing algorithm already internally computes hash-to-curve on the message so its okay to pass arbitrary messages to the signature computation algorithm and we donot need to compute a hash digest first.

isaacdoidge commented on aptos-move/aptos-vm/src/validator_txns/dkg.rs:
```diff
@@ -33,13 +28,14 @@ use move_vm_types::gas::UnmeteredGasMeter;
 enum ExpectedFailure {
     // Move equivalent: `errors::invalid_argument(*)`
     EpochNotCurrent = 0x10001,
-    TranscriptDeserializationFailed = 0x10002,
```
Is deserialization no longer necessary in the Move layer?

hsaleemsupra replied:
We do not need to deserialize the transcript as it was only done by the aptos dkg. For our dkg, we only need to validate the aggregate signature on the transcript and record the raw transcript on chain so it can be consumed by the validators later.

isaacdoidge commented on aptos-move/framework/supra-framework/sources/dkg.move:
```diff
@@ -84,16 +120,45 @@ module supra_framework::dkg {
+        session.target_committees_public_key_shares = target_committees_public_key_shares;
         dkg_state.last_completed = option::some(session);
         dkg_state.in_progress = option::none();
+
```
From our previous discussion, I think you're already clear on this, but this is where we should propagate the new public key shares to `stake.move`.

isaacdoidge commented on aptos-move/framework/supra-framework/sources/dkg.move:
```diff
@@ -84,16 +120,45 @@ module supra_framework::dkg {
+        // we only add the first DKG Meta proposed and ignore the rest
         let session = option::extract(&mut dkg_state.in_progress);
-        session.transcript = transcript;
+        assert!(vector::length(&session.dkg_meta_transcript) == 0, error::already_exists(EDKG_META_ALREADY_SET));
```
Would it make sense to use an option here instead of a vector? Will the vec ever contain more than one element?

hsaleemsupra replied:
I am using a byte vector (`vector<u8>`) to represent the raw `dkg_meta_transcript`. Here empty vector means the dkg meta is not set and a non-empty vector means its already set. We could also use `option<vector<u8>>` but I do not see any additional benefit of doing that.

isaacdoidge commented on aptos-move/framework/supra-framework/sources/dkg.move:
```diff
@@ -53,29 +79,39 @@ module supra_framework::dkg {
+            move_to<DKGResharing>(
+                supra_framework,
+                DKGResharing {
+                    is_resharing: false,
```
I don't see a function for setting this flag yet, are you planning to add that later?

hsaleemsupra replied:
I will add a function to set this.

isaacdoidge commented on types/src/validator_public_keys.rs:
The Move file looks good now, just need to update this one to match.

hsaleemsupra replied:
fixed

isaacdoidge commented on types/src/dkg/transactions.rs:
```diff
@@ -0,0 +1,74 @@
+pub struct DKGTransactionMetadata {
+    pub epoch: u64,
+    pub author: AccountAddress,
+    pub bls_aggregate_signature: Vec<u8>,
```
Does this signature include the chain id?

hsaleemsupra replied:
Currently it does not. I will see how to incorporate that.

hsaleemsupra replied:
chain_id has been added.

isaacdoidge commented on types/src/dkg/dkg_committee.rs:
```diff
@@ -0,0 +1,24 @@
+/// Reflection of `0x1::dkg_committee::DkgNodeConfig` in rust.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct DkgNodeConfig {
+    pub pool_address: AccountAddress,
```
These structs look like they should also be read-only. Please use the `Getters` macro if they don't need to be written to after construction.

isaacdoidge commented on aptos-move/framework/supra-framework/sources/validator_public_keys.move:
```diff
@@ -0,0 +1,223 @@
+                = option::extract(&mut validator_public_keys.supra_keys.bls_threshold_quorum_certificate_key);
+            let valid_bls_threshold_quorum_key
+                = bls12381::public_key_from_bytes(
+                bls12381::public_key_to_bytes (&bls_threshold_quorum_key));
```
There are a few formatting issues in this file. Please double check it.
```suggestion
                bls12381::public_key_to_bytes(&bls_threshold_quorum_key));
```

isaacdoidge commented on aptos-move/framework/supra-framework/sources/validator_public_keys.move:
```diff
@@ -0,0 +1,223 @@
+
+        // validate network ed25519 key
+        let valid_network_key
+            = ed25519::new_validated_public_key_from_bytes(
```
Why are we serializing then deserializing each key?

hsaleemsupra replied:
The deserialization is actually: deserialization + key validation. It is important to check that each key is a valid key. The deserialization function has checks like: key has valid length, key is a valid point on the curve, and in case of class group key it also validates proof of possession.

isaacdoidge replied:
Are we not doing this when we first deserialize the `ValidatorPublicKeys` struct?

hsaleemsupra replied:
Yeah I think, that makes sense. I will remove the validation.

isaacdoidge commented on aptos-move/framework/supra-framework/sources/dkg_committee.move:
```diff
@@ -0,0 +1,106 @@
+        }
+    }
+
+    public fun new_dkg_committee_from_validator_consensus_info(type: DkgCommitteeType, validator_committee: vector<ValidatorConsensusInfo>): DkgCommittee{
```
Please run the formatter on this file too.

isaacdoidge commented on aptos-move/framework/supra-framework/sources/dkg_committee.move:
```diff
@@ -0,0 +1,106 @@
+
+                vector::push_back(&mut dkg_committee, DkgNodeConfig{
+                    addr: validator_consensus_info::get_addr(&x),
+                    identity: ed_key_bytes,
```
We've just updated the node`Identity` to its delegation pool address for the rest of the L1 code. Can we do the same here?

hsaleemsupra replied:
Can you please explain how to access delegation pool address here? Also is there a way to serialize `address` since on moonshot side, Identity is a 32 byte array?

isaacdoidge replied:
The delegation pool address is the `addr` field of the validator consensus info. You're already using it as the `addr` field of the `DkgNodeConfig`. The `Identity` from `smr-moonshot` is just the bytes of the `AccountAddress` (i.e. [self.0](https://github.com/Entropy-Foundation/aptos-core/blob/dev/third_party/move/move-core/types/src/account_address.rs#L145)).

hsaleemsupra replied:
fixed, please check

hsaleemsupra committed:
ongoing changes to address comments

isaacdoidge commented on aptos-move/aptos-gas-schedule/src/gas_schedule/supra_stdlib.rs:
```diff
@@ -0,0 +1,21 @@
+        // Note(Gas): this initial value is guesswork.
+        [class_groups_per_pubkey_deserialize: InternalGasPerArg, { RELEASE_V1_16_SUPRA_V1_8_0.. => "class.groups.per_pubkey_deserialize" }, 400684],
+        // Note(Gas): this initial value is guesswork.
+        [class_groups_pop: InternalGas, "class.groups.base", 206000000],
```
Isn't this new too? It should also be versioned.

hsaleemsupra committed:
ongoing changes

hsaleemsupra committed:
Merge branch 'dev' into feature/dkg-integration

hsaleemsupra committed:
minor fixes

isaacdoidge commented on aptos-move/framework/supra-framework/sources/validator_public_keys.move:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    public fun validator_public_keys_from_bytes(bytes: vector<u8>): ValidatorPublicKeys{
+
```
```suggestion
```

isaacdoidge commented on types/src/dkg/dkg_committee.rs:
```diff
@@ -0,0 +1,25 @@
+
+/// Reflection of `0x1::dkg_committee::DkgCommittee` in rust.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Getters)]
+pub struct DkgCommittee {
```
Is this struct stored as a Resource in the Move state, or as a part of another Resource?

isaacdoidge commented on types/src/dkg/transactions.rs:
```diff
@@ -0,0 +1,77 @@
+pub struct DKGTransactionMetadata {
+    epoch: u64,
+    author: AccountAddress,
+    bls_aggregate_signature: Vec<u8>,
```
I think I might have mentioned this in a previous review, but this also needs to cover the chain id.

hsaleemsupra replied:
The data_bytes now include `EpochId (epoch, chain_id)` and the aggregate signature is computed on that. Please refer to [this](https://github.com/Entropy-Foundation/smr-moonshot/blob/643e7e491362c7a50f203fa8ce74cba2258ea2d7/crates/libs/dkg_types/src/dkg_meta.rs#L154).

hsaleemsupra committed:
adds new method for ValidatorPublicKeys

hsaleemsupra committed:
new methods for DkgNodeConfig and DkgCommittee

hsaleemsupra committed:
update dkg committee to use validator address as identity

hsaleemsupra committed:
add copyright statements

isaacdoidge commented on types/src/validator_public_keys.rs:
We'll need to update this file and the related logic after https://github.com/Entropy-Foundation/aptos-core/pull/305 has been merged.

isaacdoidge commented on aptos-move/framework/supra-stdlib/sources/class_groups.move:
```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2024 Supra.
```
```suggestion
// Copyright © 2025 Supra.
```

isaacdoidge commented on .cursorrules:
Please add this to `.gitignore` for now and remove it from the remote. We can discuss adding a ruleset like this with the rest of the team.

so-schen committed:
Sc/task/issue 2242 patch2 (#306)

* add supra bcft cert feature flag

* update md file

* addressing comments

so-schen committed:
Update validator_public_keys.move (#305)

* Update validator_public_keys.move

* Add functions for new CertificateThresholdType constant tags (BCFT and Clan Majority)

* addressing comments

* rename

* fix docs

mzajac-supra committed:
made compatible with the new key format

mzajac-supra committed:
added a fallback to the old format

mzajac-supra committed:
parsing ed key by `try_from` instead of bcs

hsaleemsupra committed:
minor fix

hsaleemsupra committed:
resharing config changed to per receiver committee

hsaleemsupra committed:
add epoch, chain_id in onchainaggcommitments

hsaleemsupra committed:
update validator committee keys after dkg

hsaleemsupra commented on aptos-move/framework/supra-framework/sources/stake.move:
```diff
@@ -847,6 +850,56 @@ module supra_framework::stake {
         );
     }
 
+    public(friend) fun set_dkg_output_keys(
```
@isaacdoidge please verify the correctness of this function.

isaacdoidge replied:
Looks good, but I think we could make this function and the related logic more robust against ordering changes in functions like `next_validator_consensus_infos` if we sort/index keys by validator index.

hsaleemsupra committed:
fix test public key

hsaleemsupra committed:
changes to support arbitrary threshold types

hsaleemsupra committed:
dkg for validity & quorum keys

hsaleemsupra committed:
Merge remote-tracking branch 'origin/dev' into feature/dkg-integration

hsaleemsupra committed:
Merge branch 'smr-moonshot-2300-consensus-keys-relaxation' into feature/dkg-integration

hsaleemsupra committed:
Update dkg.md

hsaleemsupra committed:
fixes

hsaleemsupra committed:
conditional consensus key based on bls flag

hsaleemsupra committed:
update validator_config to use ValidatorPublicKeys

hsaleemsupra committed:
update ValidatorPublicKeys to include bcft keys

hsaleemsupra committed:
update gas ver

hsaleemsupra committed:
add supra_stdlib gas params to_on_chain_gas_schedule

hsaleemsupra committed:
add randomness tag to block_prologue_ext

hsaleemsupra committed:
temporarily mark randomness biasable

hsaleemsupra committed:
add SUPRA_DKG to default features

hsaleemsupra committed:
dkg module fixes

hsaleemsupra committed:
fix dkg transactions error handling in vm

hsaleemsupra committed:
added dkg config

hsaleemsupra committed:
dkg handles all bls key types

hsaleemsupra committed:
remove dkg meta set condition for pk shares

hsaleemsupra committed:
adds dkg threshold type for receiver committees

hsaleemsupra committed:
removed error status DKG_META_NOT_SET

hsaleemsupra committed:
distinguish error code for future/past epochs

dhaval-supraoracles committed:
Merge from dev

dhaval-supraoracles committed:
fixed leader ban test cases

hsaleemsupra committed:
dkg for bcft and clan majority

hsaleemsupra committed:
update crypto version

hsaleemsupra committed:
adds next_epoch_validator_consensus_infos_for_dkg

vpanchal-supra committed:
feat: changes related to the transaction inclusion proof feature (#341)

* feat(types,accumulator): Add features for transactions inclusions and events emissions proofs.

- RLP's `Encodable` trait implementation for Contract Events.
- `get_proof_by_position` method for Merkle Accumulator.

* feat: add `EventV2` api type

* chore: cargo feature optimization

* refactor(aptos-api-types): transformed `Event` to `EventV1` and applied Full-Clone on `EventV2`

* feat: add `hash` in `EventV1` with serialization disabled

* docs: fix oai docs for `EventV1`

* feat(aptos_api_types): add `try_into_v2_events` to convert `ContractEvent` to `EventV2`

* feat: add `SUPRA_TRANSACTIONS_INCLUSION_PROOFS` feature flag

* feat: replace rlp encoding with custom encoding to generate event hash

* chore: replace `alloy` with `sha3` for `Keccak256` impl

* fix: Keccack256 hash implementation using `sha3`

---------

Co-authored-by: Nikita Puzankov <n.puzankov@supraoracles.com>
Co-authored-by: Simon Chen <s.chen@supraoracles.com>

vpanchal-supra committed:
fix: ensure `hash_bytes` gets updated with finalized hash

isaacdoidge committed:
Updated crypto ref to revision containing ref to bicycl-rs with fully-qualified SSH submodule refs.

hsaleemsupra committed:
update crypto

isaacdoidge committed:
Updated minimum GAS_UNIT_PRICE from 100 to 100_000 in accordance with the outcome of https://vote.supra.com/proposal/3. (#344)

isaacdoidge commented on aptos-move/aptos-vm/src/validator_txns/dkg.rs:
```diff
@@ -5,131 +5,108 @@ use crate::{
-
 enum ExecutionFailure {
-    Expected(ExpectedFailure),
+    Expected(DiscardedVMStatus),
```
It seems that the previous expected errors were more precise. Is there any risk in categorizing errors incorrectly?

hsaleemsupra replied:
The `DiscardedVMStatus` has various specific `StatusCode`'s now for different DKG related errors and we process them using these codes on the moonshot side. If you see `validate_dkg_validator_transaction` we still have very specific codes for various errors but this change makes it easy to process these codes both in aptos-core and moonshot.

isaacdoidge commented on aptos-move/aptos-vm/src/validator_txns/dkg.rs:
```diff
@@ -5,131 +5,108 @@ use crate::{
-            })
-            .map_err(|r| Unexpected(r.unwrap_err()))?;
+            .map_err(|e| expect_only_successful_execution(e, function_name.as_str(), log_context))
+            .map_err(|r| ExecutionFailure::Unexpected(r.unwrap_err()))?;
```
What does the executor do when it receives an unexpected error? We need to ensure that malicious nodes cannot cause their peers to crash by sending invalid inputs.

hamzahsaleem replied:
For `Expected` errors, we process the `StatusCode` and based on that decide whether to retry the transaction or ignore it. In case a malicious node submits a dkg transaction with invalid inputs, that should cause an `Expected` error as we validate the dkg transaction in `validate_dkg_validator_transaction` and any transaction with invalid inputs should be caught there. So in this case, we do not expect a panic or crash, rather the error is appropriately handled.

The `Unexpected` error is related to some unexpected issue in the VM while calling the move functions. For this error, the executor with crash based on the following flow:

In execute_move_transactions_internal (`core.rs:105-107)`:

 ```
 let outputs = move_executor
      .execute_transaction_block_parallel(&move_store, move_transactions.clone())
      .map_err(|vm_status| ExecutionError::from((vm_status, context)))?;
```

 This calls `BlockAptosVM::execute_block` with the entire set of transactions — user transactions, block metadata, and the DKG validator transaction together.
 
 This propagates via ? up through `execute_move_transactions` → `execute_transactions` → `execute_and_apply_certified_block` → `execute_block`.

At `consensus/moonshot/src/core.rs:1087-1097`:

 ```
 self.executor
      .execute_block(&block, validator_committee, failed_leader_indices)
      .await
      .unwrap_or_else(|e| {
          panic!(
              "Fatal: Executor terminated with error {e} while executing \
              commit-certified block {block:?}"
          )
      });

```
  The process panics.

isaacdoidge replied:
Thanks for the detailed explanation. Yes, unexpected internal errors should cause panics. We just don't want malicious user inputs to be able to trigger them as a part of the normal protocol. Everything seems fine here.

hsaleemsupra committed:
Merge branch 'dev' into feature/dkg-integration

hsaleemsupra committed:
test fixes

isaacdoidge commented on aptos-move/framework/move-stdlib/sources/configs/features.move:
```diff
@@ -784,6 +784,59 @@ module std::features {
+        is_enabled(SUPRA_BCFT_CERTIFICATES)
+    }
+
+    /// Whether the APIs related to the Supra_Dkg feature are enabled.
```
Need to note that this feature depends on the `SUPRA_BLS_KEYS` feature. I'll add a comment.

hsaleemsupra committed:
make order of dkg committee consistent with new epoch committee

isaacdoidge committed:
Fixed DKG config to account for BCFT flag and added missing threshold. Added comments. Regenerated docs.

isaacdoidge committed:
Fixed DKG config and removed unnecessary test.

isaacdoidge committed:
Discovered that DKG config is initialised during genesis before feature flags are set, so updated implementation accordingly and added notes. Also fixed supra_governance::reconfigure.

isaacdoidge committed:
Formatting.

isaacdoidge committed:
Started work towards adding leader ban registry config to genesis. (#346)

* Started work towards adding leader ban registry config to genesis.

* added remaining changes for the leader ban config at genesis

* added deserilization step

* added deserilization step

* added deserilization step

* fixed leader ban registry package for ban parameters

* decoded correct version of params

---------

Co-authored-by: Dhaval Purohit <d.purohit@supraoracles.com>

isaacdoidge commented on aptos-move/aptos-vm/src/aptos_vm.rs:
```diff
@@ -2345,11 +2358,17 @@ impl AptosVM {
+        session
+            .get_native_extensions()
+            .get_mut::<RandomnessContext>()
+            .mark_unbiasable();
```
We need to gate this with the DKG feature flag and also add a comment explaining that the randomness is still biasable during the first epoch in which the feature is activated because threshold keys are not available until the end of that epoch.

isaacdoidge commented on aptos-move/framework/supra-framework/sources/dkg_committee.move:
```diff
@@ -0,0 +1,123 @@
+    ): DkgCommittee {
+        assert!(vector::length(&validator_committee) > 0, EINVALID_DKG_COMMITTEE_SIZE);
+
+        // The order of the committee members is important for DKG.
+        // The order should correspond to the order of the validator committee.
+        // The output of the DKG has keys in the same order as the committee.
+        let dkg_committee = vector[];
+        vector::for_each(
+            validator_committee,
+            |x| {
+                let validator_keys_bytes = validator_consensus_info::get_pk_bytes(&x);
+                let addr = validator_consensus_info::get_addr(&x);
+                vector::push_back(
+                    &mut dkg_committee,
+                    DkgNodeConfig {
+                        addr,
+                        identity: bcs::to_bytes(&addr),
+                        dkg_pubkey: validator_keys_bytes
+                    }
+                );
+            }
+        );
```
Seems like it might be more robust to sort by node index.

isaacdoidge replied:
This seems to be working well in `smr-moonshot` now, so this change probably isn't necessary.

isaacdoidge commented on aptos-move/framework/supra-framework/sources/stake.move:
```diff
@@ -1467,56 +1711,216 @@ module supra_framework::stake {
+    /// Same as `next_validator_consensus_infos` but uses the current-epoch config data for
+    /// validators that are already active. This is used when starting DKG: the committee
+    /// membership (active + pending_active) must reflect the next epoch, but node config should reflect
+    /// current-epoch for the active validators.
```
Is this elaboration correct?
```suggestion
    /// Same as `next_validator_consensus_infos` but uses the current-epoch config data for
    /// validators that are already active. This is used when starting DKG: the committee 
    /// membership (active + pending_active) must reflect the next epoch, but node config
    /// should reflect current-epoch for the active validators to ensure that validators that update
    /// their configuration do not need to run two copies of themselves to participate in DKG.
```

hamzahsaleem replied:
yes looks good

isaacdoidge commented on aptos-move/framework/supra-framework/sources/stake.move:
```diff
@@ -1467,56 +1711,216 @@ module supra_framework::stake {
+                    // the order. Mirror that here so validator indices match after epoch transition.
+                    let pending_idx =
+                        num_cur_pending_actives - 1 - (candidate_idx - num_cur_actives);
+                    vector::borrow(&cur_validator_set.pending_active, pending_idx)
```
Why do we not need to do this in `next_validator_consensus_infos`?

hamzahsaleem replied:
For consistency, I have reversed the order for pending active in both `next_validator_consensus_infos` and `next_validator_consensus_infos_for_dkg`. Also there was a bug in `set_dkg_output_keys` which was calling `next_validator_consensus_infos` while the dkg was run with the reversed order of pending active nodes. So making the order consistent in both `next_validator_consensus_infos` and `next_validator_consensus_infos_for_dkg` fixes this bug.

isaacdoidge replied:
Looks great, thank you.

isaacdoidge commented on aptos-move/framework/supra-framework/sources/stake.move:
```diff
@@ -1467,56 +1711,216 @@ module supra_framework::stake {
+    /// validators that are already active. This is used when starting DKG: the committee
+    /// membership (active + pending_active) must reflect the next epoch, but node config should reflect
+    /// current-epoch for the active validators.
+    public fun next_epoch_validator_consensus_infos_for_dkg(): vector<ValidatorConsensusInfo> acquires ValidatorSet, ValidatorPerformance, StakePool, ValidatorFees, ValidatorConfig {
```
It seems like this function shares a lot of code with `next_validator_consensus_infos`. Is it possible to abstract the common functionality into new functions?

isaacdoidge commented on rust-toolchain.toml:
```diff
@@ -1,5 +1,5 @@
 [toolchain]
-channel = "1.78.0"
+channel = "1.89.0"
```
Let's update to 1.93 if we can do so easily, since that's what we're moving to in `smr-moonshot`.

isaacdoidge committed:
Formatting.

isaacdoidge committed:
Added TypeTags for UpdateNetworkAndFullnodeAddressesEvent and RotateConsensusKeyEvent.

dhaval-supraoracles committed:
Added fixes for leader ban registry view (#350)

* added fixes for leader ban registry view

* fixed the test according to updated changes

* PR comments addressed

hsaleemsupra committed:
addressing comments

hsaleemsupra committed:
reverse pending active order for next_validator_consensus_infos
